src/Control/Concurrent/CachedIO.hs

```diff
@@ -39,9 +53,23 @@ cachedIOWith
     => (UTCTime -> UTCTime -> Bool) -- ^ Test function:
     --   If 'isCacheStillFresh' 'lastUpdated' 'now' returns 'True'
     --   the cache is considered still fresh and returns the cached IO action
-    -> t a                          -- ^ action to cache
+    -> t a -- ^ action to cache.
+    -> m (t a)
+cachedIOWith f io = cachedIOWith' f (const io)
+
+-- | Cache an IO action, The cache begins uninitialized.
+--
+-- The outer IO is responsible for setting up the cache. Use the inner one to
+-- either get the cached value or refresh
+cachedIOWith'
+    :: (MonadIO m, MonadIO t, MonadCatch t)
+    => (UTCTime -> UTCTime -> Bool) -- ^ Test function:
+    --   If 'isCacheStillFresh' 'lastUpdated' 'now' returns 'True'
```

Comment:
Single quotes are for linkified identifiers, I would describe the arguments positionally.